2020/12/rain/src/main.rs

```diff
@@ -0,0 +1,440 @@
+use std::env;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+
+fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+    let mut reader = BufReader::new(File::open(&args[1])?);
+    let script = Action::parse_script_from_bufread(&mut reader)?;
+
+    let mut ship = Ship::new();
+    for action in script {
+        ship = ship.apply_part2(&action);
+    }
+
+    println!(
+        "The ship is at a manhattan distance of {}",
+        ship.manhattan_distance()
+    );
+
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum Error {
+    IO(std::io::Error),
+    Parse(std::num::ParseIntError),
+    TooShort,
+    WrongDirection(char),
+}
+
+impl From<std::io::Error> for Error {
+    fn from(ioe: std::io::Error) -> Self {
+        Error::IO(ioe)
+    }
+}
+
+impl From<std::num::ParseIntError> for Error {
+    fn from(pie: std::num::ParseIntError) -> Self {
+        Error::Parse(pie)
+    }
+}
+
+pub enum Action {
+    North(isize),
+    South(isize),
+    East(isize),
+    West(isize),
+    Left(isize),
+    Right(isize),
+    Forward(isize),
+}
+
+impl Action {
+    pub fn parse_script_from_string(script: &str) -> Result<Vec<Self>, Error> {
+        let mut res = Vec::new();
+        for line in script.lines() {
+            res.push(Action::parse(line)?);
+        }
+        Ok(res)
```

Comment:
I think `Cursor<T: AsRef<[u8]>>` implements `BufRead` so any type that can be referenced as a slice of bytes can be wrapped in a `Cursor` and passed in. So I'm pretty sure this will work:

```rust
parse_script_from_bufread(Cursor::new(script))
```

A `Cursor` is really handy for turning byte data into a read stream. Basically it's the `MemoryStream` of rust :).

Reply:
That's what I've been looking for.  My google-fu has been failing to find that.

Reply:
Yeah, `Cursor` is quite clever but not the most obvious thing to look for